### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.2663397789001465 seconds
257 ms ± 2.6 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [1]:
# 3.2 a)
import time

# Crear una lista de Python con 10^6 elementos (1,000,000)
my_list = list(range(1000000))

# Opción 1: Sumar con un bucle for
def for_loop_sum():
    total = 0
    for num in my_list:
        total += num
    return total

# Medición del tiempo para el bucle for
initialTime = time.time()
total_for = for_loop_sum()
finalTime = time.time()
print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Opción 2: Sumar usando la función sum()
def sum_function():
    return sum(my_list)

# Medición del tiempo para sum()
initialTime = time.time()
total_sum = sum_function()
finalTime = time.time()
print("Time taken by sum() operation:", (finalTime - initialTime), "seconds")

# Medición con %timeit
print("\nTiming with %timeit:")

# Medición con %timeit para el bucle for
%timeit -r 2 -n 1000 for_loop_sum()

# Medición con %timeit para la función sum()
%timeit -r 2 sum_function()

# Mostrar el resultado de la suma
print(f"\n\tComputing the sum of numbers in the range [0, 1000000]: {total_sum}\n")
import time

# Crear una lista de Python con 10^6 elementos
my_list = list(range(1000000))

# Opción 1: Sumar con un bucle for
def for_loop_sum():
    total = 0
    for num in my_list:
        total += num
    return total

# Opción 2: Sumar usando sum() de Python
def sum_function():
    return sum(my_list)

# Medición con time.time() para el bucle for
initialTime = time.time()
total_for = for_loop_sum()
finalTime = time.time()
print("Time taken by reduction operation (for loop):", (finalTime - initialTime), "seconds")

# Medición con time.time() para sum()
initialTime = time.time()
total_sum = sum_function()
finalTime = time.time()
print("Time taken by sum() operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
print("\nTiming with %timeit:")

# Medición con %timeit para el bucle for
%timeit -r 2 -n 1000 for_loop_sum()

# Medición con %timeit para la función sum()
%timeit -r 2 sum_function()

# Mostrar el resultado de las sumas de todas las maneras posibles
print(f"\n\tComputing the sum of numbers in the range [0, 1000000]:")
print(f"\tSum using for loop: {total_for}")
print(f"\tSum using sum(): {total_sum}")

Time taken by reduction operation: 0.23157072067260742 seconds
Time taken by sum() operation: 0.03262734413146973 seconds

Timing with %timeit:
219 ms ± 2.11 ms per loop (mean ± std. dev. of 2 runs, 1,000 loops each)
30.5 ms ± 283 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)

	Computing the sum of numbers in the range [0, 1000000]: 499999500000



In [3]:
#3.2 b)
import time
import numpy as np

# Convertir la lista a un array de NumPy
my_array = np.array(my_list)

# Opción 1: Sumar con un bucle for
def for_loop_sum_array():
    total = 0
    for num in my_array:
        total += num
    return total

# Medición con time.time() para el bucle for
initialTime = time.time()
total_array_for = for_loop_sum_array()
finalTime = time.time()
print("Time taken by reduction operation with array:", (finalTime - initialTime), "seconds")

# Opción 2: Sumar usando np.sum() de NumPy
def sum_function_array():
    return np.sum(my_array)

# Medición con time.time() para np.sum()
initialTime = time.time()
total_array_sum = sum_function_array()
finalTime = time.time()
print("Time taken by np.sum() operation with array:", (finalTime - initialTime), "seconds")

# Medición con %timeit
print("\nTiming with %timeit:")

# Medición con %timeit para el bucle for
%timeit -r 2 -n 1000 for_loop_sum_array()

# Medición con %timeit para la función np.sum()
%timeit -r 2 sum_function_array()


Time taken by reduction operation with array: 0.6622755527496338 seconds
Time taken by np.sum() operation with array: 0.0020012855529785156 seconds

Timing with %timeit:
648 ms ± 6.19 ms per loop (mean ± std. dev. of 2 runs, 1,000 loops each)
1.43 ms ± 234 ns per loop (mean ± std. dev. of 2 runs, 1,000 loops each)


Los resultados muestran que el bucle for es el enfoque más lento para realizar la suma, lo que refleja la sobrecarga de Python al manejar operaciones repetitivas de manera explícita. Aunque este método es directo, su rendimiento es limitado cuando se trabaja con grandes cantidades de datos, ya que la iteración manual no está optimizada. Por otro lado, la función sum() de Python mejora considerablemente el tiempo de ejecución gracias a su implementación optimizada en C, permitiendo una suma mucho más rápida de los elementos de la lista. Finalmente, np.sum() de NumPy ofrece un rendimiento similar al de sum(), pero su verdadera ventaja radica en su capacidad para manejar arrays más grandes y multidimensionales de forma más eficiente, aprovechando las optimizaciones propias de NumPy. Esto demuestra que usar funciones nativas optimizadas o bibliotecas especializadas como NumPy puede acelerar significativamente el procesamiento en comparación con enfoques más simples como los bucles for.

In [4]:
#3.3
import time
import numpy as np
import numba

# Crear un array de NumPy con 10^6 elementos
my_array = np.array(range(1000000))

# Opción 1: Sumar usando np.sum() de NumPy
def numpy_sum():
    return np.sum(my_array)

# Opción 2: Sumar usando Numba (con @jit)
@numba.jit
def numba_sum():
    total = 0
    for num in my_array:
        total += num
    return total

# Medición con time.time() para np.sum()
initialTime = time.time()
total_numpy = numpy_sum()
finalTime = time.time()
print("Time taken by np.sum() operation with array:", (finalTime - initialTime), "seconds")

# Medición con time.time() para Numba
initialTime = time.time()
total_numba = numba_sum()
finalTime = time.time()
print("Time taken by Numba operation with array:", (finalTime - initialTime), "seconds")

# Medición con %timeit
print("\nTiming with %timeit:")

# Medición con %timeit para np.sum()
%timeit -r 2 -n 100 numpy_sum()

# Medición con %timeit para Numba
%timeit -r 2 -n 100 numba_sum()

# Mostrar el resultado de las sumas
print(f"\n\tComputing the sum of numbers in the range [0, 1000000]:")
print(f"\tSum using np.sum(): {total_numpy}")
print(f"\tSum using Numba: {total_numba}")


Time taken by np.sum() operation with array: 0.0019183158874511719 seconds
Time taken by Numba operation with array: 3.4399187564849854 seconds

Timing with %timeit:
1.43 ms ± 2.09 µs per loop (mean ± std. dev. of 2 runs, 100 loops each)
1.69 ms ± 13.1 µs per loop (mean ± std. dev. of 2 runs, 100 loops each)

	Computing the sum of numbers in the range [0, 1000000]:
	Sum using np.sum(): 499999500000
	Sum using Numba: 499999500000


Explicación de los resultados obtenidos por Numba (3.3 b)
En este ejercicio, hemos comparado el rendimiento de NumPy y Numba al sumar los elementos de un array de 1,000,000 de elementos.

np.sum() de NumPy:
El tiempo de ejecución de np.sum() fue de 0.0019 segundos, lo que refleja su alta eficiencia debido a su implementación optimizada en C.
Al utilizar %timeit, se midió un tiempo promedio de 1.43 ms por ejecución, lo que demuestra la rapidez de NumPy en operaciones de suma.


Numba:
El tiempo de ejecución con Numba fue significativamente más largo, con 3.44 segundos en total. Este incremento se debe al tiempo de compilación JIT (Just-In-Time) que Numba realiza en la primera ejecución de la función.
Con %timeit, Numba tardó 1.69 ms por ejecución, ligeramente más que np.sum(), pero todavía rápido. La diferencia inicial se debe a la compilación que ocurre la primera vez.

Comparación de resultados:
Ambos métodos produjeron el mismo resultado de la suma, 499,999,500,000, lo que confirma que ambos enfoques suman correctamente los elementos del array.

Conclusión:
NumPy es más rápido para operaciones simples como la suma debido a su optimización nativa.
Numba, aunque es más lento en la primera ejecución por la compilación JIT, es útil para funciones más complejas y repetitivas, donde su rendimiento mejora después de la compilación inicial.


In [4]:
import argparse

# Definir el parser para los argumentos de línea de comandos
parser = argparse.ArgumentParser(description="Run reduction operation with dynamic input size.")
parser.add_argument('value', type=int, help="Number of elements for the reduction operation.")
args, unknown = parser.parse_known_args()  # Esto ignora los argumentos desconocidos

value = args.value  # Asignar el número de elementos desde la línea de comandos




usage: ipykernel_launcher.py [-h] value
ipykernel_launcher.py: error: argument value: invalid int value: '/home/alumno10/.local/share/jupyter/runtime/kernel-faaeb7d0-7d60-490b-a240-d5c773c7f150.json'


AssertionError: 

In [6]:
import argparse
import sys

def main():
    # Verificar si estamos en un entorno interactivo como Jupyter
    if hasattr(sys, 'argv') and len(sys.argv) > 1:
        # Configuración de argumentos
        parser = argparse.ArgumentParser(description="Run reduction operation with dynamic input size.")
        parser.add_argument('value', type=int, help="Number of elements for the reduction operation.")
        args, unknown = parser.parse_known_args()
        
        # Captura el valor desde los argumentos
        value = args.value
    else:
        # Definir un valor predeterminado si no se reciben argumentos
        print("No se detectaron argumentos en el entorno interactivo. Usando valor predeterminado: 100.")
        value = 100  # Valor predeterminado para entornos interactivos como Jupyter

    # Mostrar el valor y realizar la operación
    print(f"Valor recibido para la operación: {value}")
    suma = sum(range(value))
    print(f"La suma de los números en el rango [0, {value}) es: {suma}")

if __name__ == "__main__":
    main()


usage: ipykernel_launcher.py [-h] value
ipykernel_launcher.py: error: argument value: invalid int value: '/home/alumno10/.local/share/jupyter/runtime/kernel-faaeb7d0-7d60-490b-a240-d5c773c7f150.json'


AssertionError: 